statistics/stats.js

```diff
@@ -0,0 +1,298 @@
+const Jimp = require("jimp");
+const serviceAccount = require("./credentials.json");
+admin.initializeApp({
+  credential: admin.credential.cert(serviceAccount),
+  databaseURL: "https://universalgamemaker.firebaseio.com",
+  storageBucket: 'universalgamemaker.appspot.com'
+});
+
+const db = admin.database();
+const storage = admin.storage().bucket();
+const dbRef = db.ref("/gameBuilder");
+
+const db_outfile = "database.json";
+const images_outfile = "image_data.json";
+const stats_outfile = "image_stats.json";
+
+// Game name |
+// Original size in KB |
+// Size for quality=70 without compressing PNGs |
+// Size for quality=70 without compressing PNGs |
+// Size for quality=70 with compressing PNGs as JPGs |
+// Size for quality=70 with compressing PNGs as JPGs
+
+function downloadDatabase(){
+  let database_json = {};
+  dbRef.once("value", (gameBuilder) => {
+    const specs = gameBuilder.child('gameSpecs');
+    const elements = gameBuilder.child('elements');
+    const images = gameBuilder.child('images');
+
+    database_json["specs"] = specs.toJSON();
+    database_json["elements"] = elements.toJSON();
+    database_json["images"] = images.toJSON();
+    specs.forEach((spec) =>{
+      let json = spec.child('pieces').toJSON();
+      if(typeof json == "string"){ //it's a badly formatted array!
+        let array_obj = JSON.parse(database_json["specs"][spec.key]["pieces"]);
+        let build = {};
+        for(let i = 0; i < array_obj.length; i++){
+          build[i] = array_obj[i];
+        }
+        database_json["specs"][spec.key]["pieces"] = build;
+      }
+    });
+    fs.writeFileSync(db_outfile, JSON.stringify(database_json, null, 2));
+    console.log("Saved JSON to", db_outfile);
+  });
+}
+
+function downloadImages(){
+  storage.getFiles(function(err, files){
+    files.forEach((file) =>{
+      file.download({
+        destination: file.name
+      },(err) =>{
+        console.log("Downloaded file to", file.name);
+      });
+    });
+  });
+}
+
+function getGameData(){
+  let game_data = {}; // {gameName: {board:___, nonBoard: [__,___]}}
+  const string = fs.readFileSync(db_outfile);
+  const obj = JSON.parse(string);
+
+  const specs = obj['specs'];
+  const images = obj['images'];
+  const elements = obj['elements'];
+
+  for(let specId in specs) {
+    const name = specs[specId]['gameName'];
+    const imageId = specs[specId]['board']['imageId'];
+    const imagePath = images[imageId]['cloudStoragePath'];
+
+    game_data[name] = {};
+    game_data[name]['board'] = imagePath;
+    game_data[name]['nonboard'] = [];
+
+    const pieces = specs[specId]['pieces'];
+    for(let element in pieces){
+      const elementId = pieces[element]['pieceElementId'];
+      const image_set = elements[elementId]['images'];
+      for(let image in image_set){
+        const imageId = image_set[image]['imageId'];
+        const imgPath = images[imageId]['cloudStoragePath'];
+        game_data[name]['nonboard'].push(imgPath);
```

Comment:
What ensures that the img is only stored once per game?

Reply:
Same goes for board images.